src/jamf/client.ts

```
@@ -40,22 +43,150 @@ interface CreateJamfClientParams {
+
   private readonly host: string;
   private readonly username: string;
   private readonly password: string;
 
-  constructor(options: CreateJamfClientParams) {
+  private authType: AuthType;
+  private token: string;
+  private tokenExpiry: number;
+
+  private constructor(options: CreateJamfClientParams) {
     this.host = options.host;
     this.username = options.username;
     this.password = options.password;
   }
 
+  public async initialize() {
+    // if authType has value, it means it already has been initialized
+    if (this.authType) {
+      return;
+    }
+
+    const version = await this.fetchJamfProVersion();
+    if (!version) {
+      // fallback to... basic auth?
+      this.authType = AuthType.BasicAuthentication;
+      return;
+    }
+
+    if (version.major >= 10 && version.minor >= 35) {
+      this.authType = AuthType.BearerToken;
+    } else {
+      this.authType = AuthType.BasicAuthentication;
+    }
+  }
+
+  public static getInstance(options: CreateJamfClientParams) {
+    return this._instance || (this._instance = new this(options));
+  }
+
+  private async fetchJamfProVersion(): Promise<JamfProVersion | undefined> {
+    let version: string | null = null;
+
+    try {
+      // Let's try BearerToken strategy first
+      this.authType = AuthType.BearerToken;
+      const result = await this.makeRequest<any>(
+        this.getResourceUrl(`api/v1/jamf-pro-version`, false),
+        Method.GET,
+        {},
+      );
+      version = result.version;
+    } catch (err) {
+      try {
+        // if it fails, we have to try Basic Authentication one
+        this.authType = AuthType.BasicAuthentication;
+        const result = await this.makeRequest<any>(
+          this.getResourceUrl(`api/v1/jamf-pro-version`, false),
+          Method.GET,
+          {},
+        );
+        version = result.version;
+      } catch (err) {
+        // TODO: this means both attempts at getting version failed (which is unexpected)
```

I'd either throw an IntegrationProviderAPIError if this is a totally unknown error or an IntegrationProviderAuthenticationError if this is due to some kind of user input error.

---

> Also, won't this fail if they've passed in incorrect credentials?

You're right, I think this is the only way for the code to reach this catch block. 

I added an `IntegrationValidationError` for this case